[pydanticgen] Add list of lists array implementation (#1887)
linkml/linkml: docs/schemas/arrays.md

```

```{py:data} dimensions
:type: list[DimensionExpression]

Parameterization of individual dimensions (see below)
```

```{rubric} Dimensions
```

Dimensions can be further parameterized by defining `dimensions` with a list of {class}`.DimensionExpression`s, which accept:

```{py:data} alias
:type: str | None

The name of the dimension
```

```{py:data} maximum_cardinality
:type: int | None

The maximum size of this dimension (inclusive). If `None`, no maximum is set
```

```{py:data} minimum_cardinality
:type: int | None

The minimum size of this dimension (inclusive). If `None`, no minimum is set
```

```{py:data} exact_cardinality
:type: int | None

The exact size of this dimension. Equivalent to `minimum_cardinality` and `maximum_cardinality` being equal `int`s
```

(array-forms)=
### Shape Forms

The combinations of the different {class}`.ArrayExpression` properties imply four NDArray forms:

- [Any Shape](AnyShapeArrays) - Arrays without limits to their shape
- [Bounded Shape](BoundedShape) - Arrays with constraints on the number of dimensions without further parameterization
- [Parameterized Shape](ParameterizedShape) - Arrays with parameterized dimensions

> - [Complex Shape](ComplexShape) - Arrays with both numeric constraints and parameterized dimensions

Author: ```suggestion
- [Complex Shape](ComplexShape) - Arrays with both constraints on the number of dimensions and parameterized dimensions
```
I think this phrasing is clearer

Reviewer: agreed

(AnyShapeArrays)=
#### Any

An Any shaped array can take any shape. This is the simplest array form, indicating the mere presence of an array.

An any shaped array is specified with an empty `array` dictionary:

:::::{tab-set}
::::{tab-item} LinkML
:::{code-block} yaml
AnyType:
  attributes:
    array:
      range: AnyType
      array: {}

Typed:
  attributes:
    array:
      range: integer
      array: {}
:::
::::
::::{tab-item} pydantic - LoL
:::{code-block} python
class AnyType(ConfiguredBaseModel):
    array: Optional[AnyShapeArray] = Field(None)

class Typed(ConfiguredBaseModel):
    array: Optional[AnyShapeArray[int]] = Field(None)
:::
::::
::::{tab-item} numpydantic
:::{admonition} Coming Soon
Direct support for array libraries like numpy, hdf5, dask, zarr, and xarray with pydantic validation <3
::: 
::::
:::::

The resulting pydantic models use a special {class}`AnyShapeArray` class injected by pydanticgen's 

> {mod}`~linkml.generators.pydanticgen.template` system when using the List of List representation 

Author: ```suggestion
{mod}`~linkml.generators.pydanticgen.template` system when using the List of List (LoL) representation 
```

```{code-cell}
array = np.arange(4*1*2*6,dtype=int).reshape((4,1,2,6))
# array = np.ones((4,1,2,6), dtype=int)

console.print(ParameterizedArray(array=array))
```

(ComplexShape)=
#### Complex

Complex NDArrays combine all three of the prior forms.

For example:

> An array with between 5 and 7 dimensions such that...
> 
> - The first dimension has at least two items
> - The second dimension has at most five items
> - The third dimension has between two and five items
> - The fourth dimension has exactly six items

```{code-cell}
---
tags: ['hide-input']
---
sch = schemas / 'complex_shape.yaml'
render_comparison(sch, 'ComplexRangeShapeArray')
```

The only place where the syntax of complex arrays differ is that `minimum` and `maximum_number_dimensions`
are set at the length of the specified `dimensions` list by default. In order to specify an array
with a fixed number of parameterized dimensions with an arbitrary number of additional dimensions,
set `maximum_number_dimensions` to `False` :

```{code-cell}
---
tags: ['hide-input']
---
sch = schemas / 'complex_shape.yaml'
render_comparison(sch, 'ComplexAnyShapeArray')
```

## Generators

### Support

> At releast, only pydanticgen supports arrays, but arrays will be implemented gradually for the rest of the generators.

Author: ```suggestion
At release, only pydanticgen supports arrays, but arrays will be implemented gradually for the rest of the generators.
```
